### PR TITLE
win_share - Implement append paramtere for access rules

### LINKED ISSUE #59237: win_share: remove all other permissions of a share
##### SUMMARY
I have a playbook the user: App_Urb on a share EAI. Everything is ok, but If I have another user on that share, playbook remove it

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
win_share

##### ANSIBLE VERSION

```
ansible 2.8.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]

```

##### CONFIGURATION

```
[14:07:41]root@vsrvkermit playbook]# ansible-config dump --only-changed
DEFAULT_ROLES_PATH(/etc/ansible/ansible.cfg) = [u'/etc/ansible/roles', u'/datas/ansible/roles']

```

##### OS / ENVIRONMENT
Target is Windows 2012 R2

##### STEPS TO REPRODUCE

Case original:
```
C:\Windows\system32>net share EAI
Share name        EAI
Path              D:\Partage\EAI
Remark            Repertoire EAI
Maximum users     No limit
Users
Caching           Manual caching of documents
Permission        VSRVQSRM\Appli_Urba, FULL
                  PAM\EXP_TRT, READ
```
I manually added PAM\EXP_TRT with READ rules.

I run playbook like this:
```
- name: Add share EAI
  win_share:
    name: EAI
    description: Repertoire EAI
    path: D:\Partage\EAI
    list: no
    full: Appli_Urba
  when: ansible_distribution_version is version('6.2','>=')

```
This remove the created user

##### EXPECTED RESULTS

Keep original user, just add Appli_Urb if not exist

##### ACTUAL RESULTS

```
C:\Windows\system32>net share EAI
Share name        EAI
Path              D:\Partage\EAI
Remark            Repertoire EAI
Maximum users     No limit
Users
Caching           Manual caching of documents
Permission        VSRVQSRM\Appli_Urba, FULL
```

### PR DESCRIPTION
**Pull request by ShachafGoldstein**
##### SUMMARY
Fixes: #59237

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_share.py
win_share.ps1

### CODE REVIEW

File: changelogs/fragments/win_share-Implement-append-paramtere-for-access-rules.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - "win_share - Implement append paramtere for access rules (https://github.com/ansible/ansible/issues/59237)"
```
Comment (jborean93) [inline changelogs/fragments/win_share-Implement-append-paramtere-for-access-rules.yml]: parameter*

File: lib/ansible/modules/windows/win_share.ps1 [hunk 2]
```diff
@@ -47,6 +47,7 @@ $check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "b
 
 $name = Get-AnsibleParam -obj $params -name "name" -type "str" -failifempty $true
 $state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "present" -validateset "present","absent"
+$appendRules = Get-AnsibleParam -obj $params -name "appendRules" -type "bool" -default $false
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_share.ps1]: We seem to have settled on using `action` as an option name for similar modules that need to add, remove, or set from a list. I would suggest we use `rule_action` with those 3 options and default to `set` to ensure backwards compatibility. You don't need to implement the `remove` if you don't want to but having `set` and `add` should satisfy the feature request.

**Review by jborean93 — COMMENTED** [review]
I haven't gone through the code as it could potentially change based on my comments. Basically I think the interface should follow what we current do for things like `win_user_right`, win_hosts` and so on.

File: lib/ansible/modules/windows/win_share.ps1 (hunk 2 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/win_share.ps1]: done

File: lib/ansible/modules/windows/win_share.ps1 [hunk 3]
```diff
@@ -47,6 +47,7 @@ $check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "b
 
 $name = Get-AnsibleParam -obj $params -name "name" -type "str" -failifempty $true
 $state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "present" -validateset "present","absent"
+$rule_action  = Get-AnsibleParam -obj $params -name "rule_action " -type "str" -default "replace" -validateset "replace","remove","append"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_share.ps1]: I don't think it's documented but there are a few places where we have already settled on `add`, `remove`, and `set` which I think should apply here.

* https://docs.ansible.com/ansible/latest/modules/win_hosts_module.html
* https://docs.ansible.com/ansible/latest/modules/win_service_module.html
* https://docs.ansible.com/ansible/latest/modules/win_user_right_module.html

**Review by jborean93 — CHANGES_REQUESTED** [review]
I'm also looking at the module options that we have current set and I think the current layout isn't really right. Right now we have the following options to control the share permissions;

* change
* deny
* full
* read

Each option can be a comma separated string but in reality the `change`, `full, and `read` are actual access types and `deny` is the rule type itself. In reality we should have an option like `permissions` that is a list of dicts in the potentially one of the 2 formats below:

```
permissions:
- name: Everyone
  rights: full|change|read (multiple can be set)
  type: allow|deny

# or

permissions:
- name: Everyone
  full: allow
  change: deny
  read: ~  # null, key not defined, or '' will mean the right isn't set for either allow or deny
```

While the order is not important and we could use a dict where the key is the account name, having the key as the account name makes it harder for a user to quote, escape, or use a var for the value so I think the list way is a bit better.

I personally think the 2nd option is the way to go as it forces someone to set all the rights for each user under a single entry as well as forcing allow/deny and not have a mixture. As well as this it makes it easier to see what permissions are defined for each user as now they are under the same entry whereas before they were defined as a list for each individual right. What are your thoughts on this proposal?

Ultimately it would mean that we need to deprecate the `change`, `deny`, `full`, and `read` options but we can easily error out if both the older and new format is defined as well as massage the older options to the newer format in the code for easier comparisons.

**Comment by ShachafGoldstein** [comment]
@jborean93 - How should we handle the change then? I think we should maybe merge this one and ill work on a new one that refactors the module to use the new C# stuff and this change

**Comment by jborean93** [comment]
@ShachafGoldstein I was playing around with this a bit more and found out something I didn't realise before.

* Setting a `Deny` on `Full Control` automatically sets a `Deny` on the lower rights
    * Ultimately this means that any `Deny` rule means a `Deny` on the `Read` access right which is a denial of everything
* The same applies to `Allow` where setting an `Allow` on a higher right automatically applies to the lower rights

Before I thought you could set a deny on full or write but still allow a read but that doesn't seem to be the case making my previous suggestion a bit less warranted and really not needed so sorry about the confusion there.

I've pushed a change that adds the idempotent check to `rule_action: add` as before it always reported a change. I also added the `remove` action from the `Get-AnsibleParam` validateset as that is not documented or implemented yet.

**Review by jborean93 — APPROVED** [review]

**Pull request merged**